rts/motoko-rts/src/stabilization.rs

```diff
@@ -0,0 +1,490 @@
+    let serialized_data_start = stable_memory_pages * PAGE_SIZE;
+    let serialized_data_length = Serialization::run(stable_actor, serialized_data_start);
+    let type_descriptor = TypeDescriptor::new(old_candid_data, old_type_offsets, 0);
+    let metadata = StabilizationMetadata {
+        stable_memory_pages,
+        serialized_data_start,
+        serialized_data_length,
+        type_descriptor,
+    };
+    metadata.store();
+}
+
+/// Post-upgrade operation for graph-copy-based program upgrades:
+/// Deserialize the object graph stored in stable memory back into main memory by using a graph
+/// copy algorithm. Checks whether the new program version is compatible to the stored state by
+/// comparing the type tables of both the old and the new program version.
+/// The parameters encode the type table of the new program version to which that data is to be upgraded.
+/// `new_candid_data`: A blob encoding the Candid type as a table.
+/// `new_type_offsets`: A blob encoding the type offsets in the Candid type table.
+///   Type index 0 represents the stable actor object to be serialized.
+/// Returns the root object containing all restored stable variables of the actor.
+/// /// Traps if the stable state is incompatible with the new program version and the upgrade is not
+/// possible.
+/// Implementation:
+/// * Algorithm:
+///     1. Copy the serialized image from stable memory into main memory.
+///     2. Cheney's algorithm using main memory (encoded in stable layout) as from-space and
+///        stable memory (encoded in main memory layout) as to-space.
+///     3. Copy the deserialized image from stable memory back into main memory.
+///   (This multi-step approach serves for avoiding random accesses to stable memory and thus
+///    minimizing the expensive API calls on stable memory.)
+/// * Encoding: The from-space uses the stable memory layout (although located in main memory),
+///   while the to-space is to be encoded in main memory layout (although located in stable memory).
+/// * Encoding: The from-space uses the main memory heap layout, while the to-space is encoded in the
+///   stable object graph layout (see `GraphCopyStabilization.md`).
+#[ic_mem_fn(ic_only)]
+pub unsafe fn destabilize<M: Memory>(
+    mem: &mut M,
+    new_candid_data: Value,
+    new_type_offsets: Value,
+) -> Value {
+    use crate::{memory::ic::dynamic_heap_end, rts_trap_with, stable_mem::moc_stable_mem_set_size};
+    use compatibility::{memory_compatible, TypeDescriptor};
+    use metadata::StabilizationMetadata;
+
+    let mut new_type_descriptor = TypeDescriptor::new(new_candid_data, new_type_offsets, 0);
+    let metadata = StabilizationMetadata::load(mem);
+    let mut old_type_descriptor = metadata.type_descriptor;
+    if !memory_compatible(mem, &mut old_type_descriptor, &mut new_type_descriptor) {
+        rts_trap_with("Memory-incompatible program upgrade");
+    }
+    // There may already exist some objects before destabilization.
```

Comment:
Is this really true? is that because of deserialized initialization arguments?

Reply:
Yes, unfortunately, I did not yet analyze all objects, but I believe some are the actor initialization arguments, and I believe there is also the canister object itself. Some temporary objects, such as the type descriptors can be discarded, but I noticed there are other objects that need to be retained. I can analyze which objects these are such that we know better.

Reply:
That's probably possible, if anyone can still understand the somewhat convoluted code in desugar.ml.  Or maybe we should just re-arrange the backend a bit so it does the destabilization early, before any user code has a chance to execute, perhaps storing the destabilized record in some well-known global where it can be fetched later. Not sure.

Reply:
I agree, the nicer solution would be to ensure an empty heap on destabilization. I will try to rearrange the code. Perhaps, it could be quite easy, as it is only about one special case.

Reply:
**Update**: As linear scanning is not feasible with the partitioned heap (in particular due to the large object allocation), I had to change the deserialization, now using a data structure to remember the pending objects to be scanned. As a result, the deserialization does no longer make any assumptions about the heap, such that it does not need to be empty at the beginning of the deserialization process.